rpc: Add support for JSON-RPC named arguments

laanwj opened the pull request:
The [JSON-RPC specification](http://www.jsonrpc.org/specification) allows either passing parameters as an Array, for by-position arguments, or as an Object, for by-name arguments. Currently Bitcoin Core only supports by-position arguments.

This pull request makes the (minimal) changes to support by-name arguments, but preserves full backwards compatibility. APIs using by-name arguments are easier to extend - as arguments can be left out - and easier to use - no need to guess which argument goes where.

This is especially nice in languages such as Python, which have native support for named arguments. Examples from the test:

``` python
h = node.help(command='getinfo')
h = node.getblockhash(index=0)
h = node.getblock(hash=h)
```
- In this implementation named arguments are mapped to positions by a per-call structure, provided in the RPC command table. In the future calls could process the parameter object themselves, but this provides nice automatic interoperability with both worlds.
- Missing arguments will be replaced by null ("holes"), except if at the end, then the argument is left out completely.
- It is not possible to specify **both** named and positional arguments like in Python. JSON-RPC doesn't allow for this, and I think it would be only confusing anyhow.
- An error (`RPC_INVALID_PARAMETER`) will be return if an unknown named argument is specified.

Currently no calls have support for "holes" between arguments, but this should be implemented later on a per-call basis. These should be replaced with default values.
### TODO
- ~~`bitcoin-cli` should grow support for providing named arguments~~ DONE see https://github.com/bitcoin/bitcoin/pull/8811#issuecomment-262252906

- ~~Calls in `mining`, `net`, `rawtransaction` `wallet` still need argument names filled in. I'll do this if reaction to this is positive.~~

### Later

- Same for the debug console. Syntax to be determined there. Not necessarily in this pull, my focus here is the core implementation.

laanwj committed:
authproxy: Add support for RPC named arguments

laanwj committed:
rpc: Support named arguments

The [JSON-RPC specification](http://www.jsonrpc.org/specification) allows passing parameters as an Array, for by-position
arguments, or an Object, for by-name arguments.

This implements by-name arguments, but preserves full backwards compatibility. API using by-name arguments are
easier to extend, and easier to use (no need to guess which argument goes where).

Named are mapped to positions by a per-call structure, provided through the RPC command table.

Missing arguments will be replaced by null, except if at the end, then the argument is left out completely.

Currently calls fail (though not crash) on intermediate nulls, but this should be improved on a per-call basis later.

laanwj committed:
rpc: Named arguments for blockchain calls

laanwj committed:
rpc: Add 'echo' call for testing

This hidden call simply returns what is passed in.

laanwj committed:
rpc: Named arguments for misc calls

laanwj committed:
test: Add test for RPC named arguments

Add RPC testcase for RPC named arguments.

laanwj commented:
This is useful because a lot of features add new arguments to an existing call. For example #8751.

Currently the user of the API has to explicitly specify defaults for the intermediate arguments. With named arguments this would no longer be necessary.

It could also help clean up the API for calls that already have a lot of optional arguments such as `sendtoaddress`.

MarcoFalke commented on qa/rpc-tests/test_framework/authproxy.py:
```diff
@@ -132,14 +136,16 @@ def _request(self, method, path, postdata):
                                  json.dumps(args, default=EncodeDecimal, ensure_ascii=self.ensure_ascii)))
+        if args and argsn:
+            raise ValueError('Cannot handle both named and positional arguments')
         postdata = json.dumps({'version': '1.1',
```
In the OP you are linking to the "JSON-RPC 2.0 Specification", which states
>jsonrpc
    A String specifying the version of the JSON-RPC protocol. MUST be exactly "2.0".

Can you explain why you keep this at 1.1?

laanwj replied:
May make sense to change it to that (even irrespective of this pull) however I'm not sure that we support the entire 2.0 standard, so that may involve other changes too.

MarcoFalke commented:
The only alternative I see would be to only accept one object as first positional argument for all methods, but this will likely never happen, so Concept ACK.

jonasschnelli commented:
Concept ACK. I think this is very useful.
ACK to extend it to `bitcoin-cli`, to the other rpc tables and to the RPC value conversions table.

After this, we should update the documents (and RPC help examples) to slowly move away from recommending fixed position arguments.

laanwj commented:
Rebased to get rid of the univalue change

laanwj commented:
Needed rebase, rebased

instagibbs commented on src/rpc/server.cpp:
```diff
@@ -418,6 +419,47 @@ std::string JSONRPCExecBatch(const UniValue& vReq)
+    for (const std::string &argName: argNames) {
+        auto fr = argsIn.find(argName);
+        if (fr != argsIn.end()) {
+            for (int i = 0; i < hole; ++i) {
```
after this loop `hole` should be reset to 0, otherwise holes will be repeated for every subsequent entry?

laanwj replied:
Yes, good catch, thanks!

instagibbs commented on src/rpc/server.cpp:
```diff
@@ -418,6 +419,47 @@ std::string JSONRPCExecBatch(const UniValue& vReq)
+    // Process expected parameters.
+    int hole = 0;
+    for (const std::string &argName: argNames) {
+        auto fr = argsIn.find(argName);
```
what does fr stand for?

instagibbs commented on qa/rpc-tests/rpcnamedargs.py:
```diff
@@ -0,0 +1,53 @@
+        node.getblock(hash=h)
+
+        assert_equal(node.echo(), [])
+        assert_equal(node.echo(arg0=0,arg9=9), [0] + [None]*8 + [9])
```
may want to add test of [exist, hole, exist, exist] to catch gap repeat error

mrbandrews replied:
I added the test: 
assert_equal(node.echo(arg0=0,arg3=3,arg9=9), [0] + [None]*2 + [3] + [None]*5 + [9])
Verified that with the code as is it fails, but with instagibbs fix of hole=0 it succeeds.

instagibbs reviewed (changes requested):
concept ACK, and aside from the error mentioned, looks like good minimal support

afk11 commented:
Concept ACK, needs rebase

laanwj commented:
Rebased, added @instagibbs' test and fix.

mrbandrews commented:
ACK

sipa commented:
Concept ACK!

Do you intend to include the echo command in the final merge?

laanwj commented:
@sipa I've had use for it two times in testing - here and in #7783. Also I fathom it's somewhat useful for users when testing JSON RPC libraries. So my preference would be to keep it.

instagibbs commented:
utACK https://github.com/bitcoin/bitcoin/pull/8811/commits/3c7647ea7814f7d3f0304bb4b39274c748a88900

laanwj committed:
rpc: Named arguments for net calls

Also add a more descriptive message for `setnetworkactive`.

laanwj committed:
rpc: Named arguments for mining calls

laanwj committed:
rpc: Named arguments for wallet calls

laanwj committed:
rpc: Argument name consistency

The meaning is clear from the context, and we're inconsistent here.
Also save typing when using named arguments.

- `bitcoinaddress` -> `address`
- `bitcoinprivkey` -> `privkey`
- `bitcoinpubkey` -> `pubkey`

laanwj committed:
rpc: Named arguments for rawtransaction calls

laanwj commented:
As there is interest in this I'm continuing work on the issues that are still open in the OP. 
I've pushed argument naming for `mining`, `net`, `rawtransaction` and `wallet`. I'll also make it possible to use named arguments with `bitcoin-cli`.
I'll leave the GUI change for another time though.

laanwj committed:
rpc: Named argument support for bitcoin-cli

Usage e.g.:

    $ src/bitcoin-cli -testnet -named echo arg0="dfdf"
    [
    "dfdf"
    ]

Argument conversion also works, for arguments thus flagged in the table in
`src/rpc/client.cpp`.

    $ src/bitcoin-cli -testnet -named echojson arg0="[1,2,3]"
    [
      [
        1,
        2,
        3
      ]
    ]

Unknown parameter (detected server-side):

    $ src/bitcoin-cli -testnet -named getinfo arg0="dfdf"
    error code: -8
    error message:
    Unknown named parameter arg0

laanwj commented:
Added support to bitcoin-cli. Usage e.g.:

    $ src/bitcoin-cli -testnet echo --arg0="dfdf"
    [
    "dfdf"
    ]

Argument conversion also works, for arguments thus flagged in the table in
`src/rpc/client.cpp`.

    $ src/bitcoin-cli -testnet echojson --arg0="[1,2,3]"
    [
      [
        1,
        2,
        3
      ]
    ]

Unknown parameter (detected server-side):

    $ src/bitcoin-cli -testnet getinfo --arg0="dfdf"
    error code: -8
    error message:
    Unknown named parameter arg0

This should complete the feature set for this pull. I'd appreciate testing and review.

Especially important to review are the parameter names in the RPC tables, as they will become part of the API:

- https://github.com/laanwj/bitcoin/blob/2016_09_rpc_named_arguments/src/rpc/server.cpp#L275
- https://github.com/laanwj/bitcoin/blob/2016_09_rpc_named_arguments/src/rpc/blockchain.cpp#L1372
- https://github.com/laanwj/bitcoin/blob/2016_09_rpc_named_arguments/src/rpc/mining.cpp#L905
- https://github.com/laanwj/bitcoin/blob/2016_09_rpc_named_arguments/src/rpc/misc.cpp#L511
- https://github.com/laanwj/bitcoin/blob/2016_09_rpc_named_arguments/src/rpc/net.cpp#L600
- https://github.com/laanwj/bitcoin/blob/2016_09_rpc_named_arguments/src/rpc/rawtransaction.cpp#L917
- https://github.com/laanwj/bitcoin/blob/2016_09_rpc_named_arguments/src/wallet/rpcwallet.cpp#L2598

theuni commented on src/rpc/client.cpp:
```diff
@@ -171,3 +191,33 @@ UniValue RPCConvertValues(const std::string &strMethod, const std::vector<std::s
+            throw(std::runtime_error("No '=' in named argument, a value needs to be provided for every argument"));
+        }
+        std::string name = s.substr(ptr, pos-ptr);
+        std::string value = s.substr(pos+1);
```
should probably make sure that ```!name.empty() && !value.empty()``` for good measure.

laanwj replied:
In that case it should do the expected thing: send an empty name or value.
An empty value is perfectly acceptable if you want to send an empty string.
An empty name is a bit strange, agreed, but it gets caught server-side as an unknown argument.

theuni commented on src/bitcoin-cli.cpp:
```diff
@@ -288,7 +288,26 @@ int CommandLineRPC(int argc, char *argv[])
+        if (haveNamed && havePositional) {
+            throw runtime_error("cannot have both named and positional arguments, did you mean to pass arguments before the command?");
+        } else if(haveNamed) {
+            params = RPCConvertNamedValues(strMethod, args);
```
Mixing args here with GetBoolArg/GetArg is a little scary. If some command took a (for example) "-version=foo" argument, I believe it would be swallowed by line 81 instead?

theuni replied:
Ah, that's the part I was missing, thanks. Maybe a comment to that effect to remind me next time I have to ask? :)

theuni reviewed (commented):
Concept ACK, and code generally looks good. I added a few nits, most of which have actually already been addressed in subsequent commits.

At a high level, I'm worried about introducing the "txid" param name for a bunch of calls that will likely need to make the distinction between txid/wtxid soon. Wallet especially. Maybe some convention like "txhash" should be preferred instead?

luke-jr commented:
Why is there a "wtxid" ever needed as a param? "txhash" and "txid" are all I see existing at all (hash != txid).

luke-jr commented:
I think a lot of the options params ought to be flattened?

laanwj commented on src/rpc/net.cpp:
```diff
@@ -595,18 +597,18 @@ UniValue setnetworkactive(const JSONRPCRequest& request)
+    { "network",            "getconnectioncount",     &getconnectioncount,     true,  {} },
+    { "network",            "ping",                   &ping,                   true,  {} },
+    { "network",            "getpeerinfo",            &getpeerinfo,            true,  {} },
+    { "network",            "addnode",                &addnode,                true,  {"node","command"} },
```
'address' is already used for bitcoin addresses, so I'd prefer not overloading that for network addresses. Other suggestions are welcome, though

laanwj commented:
> I think a lot of the options params ought to be flattened?

Possibly. Not in this pull, though. JSON can perfectly deal with nested structures.

theuni commented:
@luke-jr Sure, txhash/txid. I was only making the point that going forward there will be a distinction between those, so it'd be good to make sure that all "txid" params here really mean txid.

luke-jr commented:
@laanwj If this pull is merged as-is, then flattening the options will involve breaking backward compatibility with this, no? ie, it may be better to just not allow for already-named options at all right now.

laanwj commented:
I pushed cfields' rename suggestions, and compiled statistics on what argument names are used in which calls. This could be helpful in determining further unifications:
```
"absolute",          // setban
"addresses",         // listunspent
"allowhighfees",     // sendrawtransaction
"amount",            // move sendfrom sendtoaddress settxfee
"amounts",           // sendmany
"bantime",           // setban
"blocks",            // getnetworkhashps
"checklevel",        // verifychain
"count",             // listtransactions
"fee_delta",         // prioritisetransaction
"from",              // listtransactions
"height",            // waitforblockheight getnetworkhashps
"include_empty",     // listreceivedbyaccount listreceivedbyaddress
"include_mempool",   // gettxout
"include_watchonly", // getbalance gettransaction listaccounts listreceivedbyaccount listreceivedbyaddress listsinceblock listtransactions
"index",             // getblockhash
"keys",              // createmultisig addmultisigaddress
"locktime",          // createrawtransaction
"maxconf",           // listunspent
"maxtries",          // generate generatetoaddress
"minconf",           // getbalance getreceivedbyaccount getreceivedbyaddress listaccounts listreceivedbyaccount listreceivedbyaddress listunspent move sendfrom sendmany
"n",                 // gettxout
"nblocks",           // generate generatetoaddress estimatefee estimatepriority estimatesmartfee estimatesmartpriority
"newsize",           // keypoolrefill
"nrequired",         // createmultisig addmultisigaddress
"numblocks",         // verifychain
"options",           // fundrawtransaction importmulti
"outputs",           // createrawtransaction
"p2sh",              // importaddress
"prevtxs",           // signrawtransaction
"priority_delta",    // prioritisetransaction
"privkeys",          // signrawtransaction
"requests",          // importmulti
"rescan",            // importprivkey importaddress importpubkey
"state",             // setnetworkactive
"subtractfeefrom",   // sendmany
"subtractfeefromamount", // sendtoaddress
"target_confirmations", // listsinceblock
"template_request",  // getblocktemplate
"timeout",           // waitfornewblock waitforblock waitforblockheight walletpassphrase
"timestamp",         // setmocktime
"transactions",      // createrawtransaction lockunspent
"txids",             // gettxoutproof
"unlock",            // lockunspent
"verbose",           // getblock getblockheader getmempoolancestors getmempooldescendants getrawmempool getrawtransaction
```
> @laanwj If this pull is merged as-is, then flattening the options will involve breaking backward compatibility with this, no? ie, it may be better to just not allow for already-named options at all right now.

If a RPC call has a designated options argument now, then a future pull - that has functionality to pass the object down to the RPC call as-is - could allow specifying those options on the top level instead. **If** we decide that flattening is desirable at all.
I don't want to handle any RPC-call-specific changes here. Mind that I don't add any "null" argument handling in RPC calls here either. This just lays the ground work.

laanwj commented:
This looks like a good candidate for unification:
```
"hexdata",           // submitblock
"hexstring",         // decoderawtransaction decodescript sendrawtransaction signrawtransaction fundrawtransaction
```

luke-jr commented:
@laanwj GBT/submitblock probably shouldn't accept named parameters at all, since they are according to BIPs...?

laanwj commented:
You don't *have* to use named arguments. For consistency I prefer all calls
to have the possibility to have their arguments named, though.

laanwj commented:
Rebased for #9144, squashed

MarcoFalke commented:
I think travis wants you to include stdexcept

jnewbery commented on src/bitcoin-cli.cpp:
```diff
@@ -286,7 +286,30 @@ int CommandLineRPC(int argc, char *argv[])
+
+        UniValue params;
+        if (haveNamed && havePositional) {
+            throw runtime_error("cannot have both named and positional arguments, did you mean to pass arguments before the command?");
```
needs to be `std::runtime_error` since we're not using namespace std

jnewbery commented on src/bitcoin-cli.cpp:
```diff
@@ -286,7 +286,30 @@ int CommandLineRPC(int argc, char *argv[])
+        bool haveNamed = false;
+        bool havePositional = false;
+        for (const std::string &s: args) {
+            if (s.size() >= 1 && IsSwitchChar(s[0]))
```
I believe this breaks passing in negative numbers as a positional argument. For example, the following cli command is valid prior to this commit:

`bitcoin-cli getnetworkhashps -1`

and returns the estimated network hashes per second in the current difficulty window. After this commit, the method fails:

```
vagrant@vagrant-ubuntu-trusty-64:~$ bitcoin-cli getnetworkhashps -1
error: No '=' in named argument, a value needs to be provided for every argument
```

I haven't checked whether there are arguments that accept negative numbers in other RPC methods.

laanwj commented:
Good point about the negative values. Hadn't thought about that. Same will
hold for string values that start with '-', such as labels. Any argument
could have been meant as literal value.

Thinking of it this could even be a security issue. Say someone controls
the first argument to a call, they could now instead provide the second by
using, say --amount=123.

 It should be made explicit instead of auto-detected. E.g.

    bitcoin-cli {method} val1 val2. ..
    bitcoin-cli -named {method} -arg1=val1 -arg2=val2 ...

On Dec 5, 2016 5:02 PM, "John Newbery" <notifications@github.com> wrote:

> *@jnewbery* commented on this pull request.
> ------------------------------
>
> In src/bitcoin-cli.cpp
> <https://github.com/bitcoin/bitcoin/pull/8811#pullrequestreview-11407412>:
>
> > @@ -286,7 +286,30 @@ int CommandLineRPC(int argc, char *argv[])
>          if (args.size() < 1)
>              throw std::runtime_error("too few parameters (need at least command)");
>          std::string strMethod = args[0];
> -        UniValue params = RPCConvertValues(strMethod, std::vector<std::string>(args.begin()+1, args.end()));
> +        args.erase(args.begin()); // Remove trailing method name from arguments vector
> +
> +        // Check for named arguments.
> +        // Note that ParseParameters stops at the first argument that doesn't
> +        // start with a switch character, which happens to be the method. This
> +        // starts scanning after the method, so arguments before and after the
> +        // command will never interfere.
> +        bool haveNamed = false;
> +        bool havePositional = false;
> +        for (const std::string &s: args) {
> +            if (s.size() >= 1 && IsSwitchChar(s[0]))
>
> I believe this breaks passing in negative numbers as a positional
> argument. For example, the following cli command is valid prior to this
> commit:
>
> bitcoin-cli getnetworkhashps -1
>
> and returns the estimated network hashes per second in the current
> difficulty window. After this commit, the method fails:
>
> vagrant@vagrant-ubuntu-trusty-64:~$ bitcoin-cli getnetworkhashps -1
> error: No '=' in named argument, a value needs to be provided for every argument
>
> I haven't checked whether there are arguments that accept negative numbers
> in other RPC methods.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/8811#pullrequestreview-11407412>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAHutuI1hM5Xg_juJObhdgU1FYkcm02hks5rFDWfgaJpZM4KF-zs>
> .
>

jnewbery commented on src/rpc/client.cpp:
```diff
@@ -20,111 +20,130 @@ class CRPCConvertParam
+    { "importmulti", 0, "requests" },
+    { "importmulti", 1, "options" },
+    { "verifychain", 0, "checklevel" },
+    { "verifychain", 1, "blocks" },
```
should be nblocks

laanwj replied:
will rename it

theuni commented:
@laanwj +1 for explicit

sipa commented:
The issue only exists for string arguments, as anything else would be invalid JSON anyway.

Also, I have a slight preference for `bitcoin-cli getblock hash=X` over `bitcoin-cli getblock -hash=X`... it feels like the latter is an option to `bitcoin-cli` rather than `getblock` (but feel free to ignore this if I'm alone with that opinion).

laanwj commented:
> The issue only exists for string arguments, as anything else would be invalid JSON anyway.

Yes, only a problem for string arguments and negative numbers. But unfortunately that's enough to make it irresponsible to 'guess'.

> Also, I have a slight preference for bitcoin-cli getblock hash=X over bitcoin-cli getblock -hash=X... it feels like the latter is an option to bitcoin-cli rather than getblock (but feel free to ignore this if I'm alone with that opinion).

I have thought about that, but didn't propose it because I think some shells interpret A=B tuples after the command as environment settings - or if not - that is a convention people assume to be so because of autoconf/make.

theuni commented:
Worth mentioning how subarg syntax is handled by other familiar progs:

git:```./bitcoin-cli -testnet -- --hash=X```
or
gdb: ```./bitcoin-cli -testnet --args --hash=X```

sipa commented:
Another idea: named arguments have to precede the command (so `bitcoin-cli -hash=X getblock`), that's never ambiguous.

laanwj commented:
> ./bitcoin-cli -testnet -- --hash=X
> ./bitcoin-cli -testnet --args --hash=X

I like supporting at least one of those. Mind that the `--`/`--args` delimiter has to be *before* the command for it to be nonambigious, otherwise it may be a legitimate string argument.

> Another idea: named arguments have to precede the command (so bitcoin-cli -hash=X getblock), that's never ambiguous.

Sure, but then they can collide with bitcoin-cli's own arguments, which we were trying to avoid at the same time. It will be tricky to figure out what arguments to send to the server and which ones to keep for ourselves, and may change per version.

But even if we manage to get that sorted out, it looks like "putting the cart before the horse" syntax-wise.

sipa commented:
> Sure, but then they can collide with bitcoin-cli's own arguments, which we were trying to avoid at the same time. It will be tricky to figure out what arguments to send to the server and which ones to keep for ourselves, and may change per version.

That can be avoided by using `hash=X` style naming ;)

> But even if we manage to get that sorted out, it looks like "putting the cart before the horse" syntax-wise.

Yes, I agree there.

jnewbery commented:
I agree with the explicit `bitcoin-cli -named {method} -arg1=val1 -arg2=val2 ...` syntax. It guarantees backwards compatibility for any applications that are calling into bitcoin-cli with the current positional argument syntax.

If we're explicitly specifying that the arguments are named, then we could even drop the dashes before the argument names, ie `bitcoin-cli -named {method} arg1=val1 arg2=val2 ...`

laanwj commented:
> If we're explicitly specifying that the arguments are named, then we could even drop the dashes before the argument names, ie bitcoin-cli -named {method} arg1=val1 arg2=val2 ...

That's what @sipa is suggesting too. Seems fine to me *if* we can be sure that there are no shells that eat these. I've only seen the syntax used for passing arguments to make/configure  and always assumed it was setting environment variables .

theuni commented:
@laanwj Not sure if you were just using configure/make as an example, but in those cases anyway, those assignments are separate from environment vars.

That's why, for ex, ```CC=clang ./configure``` != ```./configure CC=clang```. The first sets env vars that aren't necessarily persistent to the resulting builds unless the user persists them explicitly in the shell. The second allows configure to forward the vars without polluting the env. If you pass a var using the 2nd syntax, and it's not understood ("precious" in autotools-speak), it will not be forwarded. afaik, the same goes for make.

sipa commented:
@laanwj `dd` and `bitcoin-tx` also use key=value.

morcos commented:
concept ACK
I will review more thoroughly after update

I think I have a slight preference for eliminating the dashes

laanwj commented:
- Dashes have been successfully eliminated. New syntax:

    $ src/bitcoin-cli -testnet -named echo arg0="dfdf"
    [
    "dfdf"
    ]

Argument conversion also works, for arguments thus flagged in the table in
`src/rpc/client.cpp`.

    $ src/bitcoin-cli -testnet -named echojson arg0="[1,2,3]"
    [
      [
        1,
        2,
        3
      ]
    ]

Unknown parameter (detected server-side):

    $ src/bitcoin-cli -testnet -named getinfo arg0="dfdf"
    error code: -8
    error message:
    Unknown named parameter arg0

Also updated "more examples" above.

- Renamed the remaining instances of `nblocks` and made sure the naming is consistent between `client.cpp` and the appropriate dispatch table (I have a script for this, which we could make part of the travis checks at some point).
- Rebased and squashed

jnewbery committed:
Update RPC argument names

jnewbery commented:
I've spent some time looking through the RPC arguments, and my view is that the following argument names should be changed to be more descriptive:

- `getblock/index` becomes `getblock/height` to conform with `waitforblockheight` and `getnetworkhashps`
- `createrawtransaction/transactions` becomes `createrawtransaction/inputs` to conform with `createrawtransaction/outputs`
- `listtransactions/from` becomes `listtransactions/skip` to avoid confusion with 'from' as in 'from address...'
- `getblock/hash` becomes `getblock/blockhash` ...
- `getblockheader/hash` becomes `getblockheader/blockhash` ..
- `preciousblock/hash` becomes `preciousblock/blockhash` ...
- `invalidateblock/hash` becomes `invalidateblock/blockhash` ...
- `reconsiderblock/hash` becomes `reconsiderblock/blockhash` ...
- ... and `waitforblock/hash` becomes `reconsiderblock/blockhash` to confirm with `gettxoutproof` `listsinceblock` and to avoid confusion with other hash types

There's a patch set which makes these changes (as well as generally cleaning up RPC comments) at https://github.com/jnewbery/bitcoin/commit/f08b91c21b9e715ec273cf431d2ec2bb8916def7

laanwj commented:
@jnewbery Sounds good to me, thanks, will cherry-pick that.

jnewbery commented:
Oops - I didn't change the named arguments in rpcnamedargs.py to match the new argument names.

Fix is here: https://github.com/jnewbery/bitcoin/commit/d45ef7c02cf682f59f129b2af5df064270b20e03

laanwj commented:
@jnewbery Thanks - cherry-picked and squashed into the last commit.

laanwj commented:
Rebased for a small add/add conflict in `rpc-tests.py`.

laanwj commented:
Rebased for minor argument documentation conflict.

sipa commented:
utACK 65bf5d27c578e7560128a28af08fef57135ec841.

I have not read through the test code in detail, nor verified whether the RPC help names match the named arguments in the server and client tables. I have also not checked the consistency of the chosen names (but I don't care strongly). I have played around briefly with the new `echo` and `echojson` commands and they seem to work as expected.

TheBlueMatt commented on src/rpc/client.cpp:
```diff
@@ -171,3 +191,31 @@ UniValue RPCConvertValues(const std::string &strMethod, const std::vector<std::s
+        std::string name = s.substr(0, pos);
+        std::string value = s.substr(pos+1);
+
+        // allow include-mempool=true as well as include_mempool=true
```
Why?

laanwj replied:
This was useful when the syntax was -arg=value. I guess it no longer really is now that plain name=value pairs are used.

TheBlueMatt reviewed (commented):
Didn't bother to audit that help text matches the server and client tables, but glanced at a few and they look good. None of the comments really matter all that much.

utACK 65bf5d27c578e7560128a28af08fef57135ec841

laanwj commented on src/rpc/server.cpp:
```diff
@@ -420,6 +421,48 @@ std::string JSONRPCExecBatch(const UniValue& vReq)
+    const std::vector<UniValue>& values = in.params.getValues();
+    std::unordered_map<std::string, const UniValue*> argsIn;
+    for (size_t i=0; i<keys.size(); ++i) {
+        argsIn[keys[i]] = &values[i];
```
Almost no JSON implementation checks for duplicate keys. The usual behavior is that the last value assigned to the key takes precedence. I think that will happen here?

TheBlueMatt replied:
Yes, I believe the last one will take precedence. I'm ok with not fixing this, but it seems super strange.

laanwj replied:
Well if it's any guide, python's JSON behaves exactly the same:
```
>>> json.loads('{"a":"b","a":"C"}')
{'a': 'C'}
```
What I would not mind is checking at the client side for duplicate arguments, as a user friendly feature, but I'm not going to do it in this pull.

laanwj commented on src/rpc/blockchain.cpp:
```diff
@@ -279,7 +279,7 @@ UniValue waitforblockheight(const JSONRPCRequest& request)
             "\nReturns the current block on timeout or exit.\n"
             "\nArguments:\n"
-            "1. block height to wait for (int)\n"
+            "1. height  (required, int) Block height to wait for (int)\n"
```
OK, will make that consistent.

laanwj commented:
Fixed @bluematt's documentation nit and removed the `_`/`-` interchangeability for key names, and squashed. Should be ready for merge now if travis passes.

jtimon commented:
Concept ACK (I'm surprised I didn't do it before).
